src/core/umount.c

```diff
@@ -381,6 +383,112 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int wait_for_child(pid_t pid) {
+        sigset_t mask;
+        int r, status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + DEFAULT_TIMEOUT_USEC;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitpid(pid, &status, WNOHANG) == pid)
+                                        /* Success */
+                                        return (WEXITSTATUS(status) == 0 ? 0 : -WEXITSTATUS(status));
```

Comment:
this doesn't look right, the child might also have exited abnormally... i.e. WIFEXITED() might not actually be true. There should probably be some way to handle that. Also in your current implementation the child processes exit with 0 or 1, and thus this function will return 0 or -1, but the latter is not a proper errno... We usually try to return negative `-errno`-style error codes, see CODING_STYLE, and we should do the same here.

It mght make sense to copy the behaviour of wait_for_terminate_and_warn() for this, i.e. return -EPROTO for abnormal process exits, and the positive exit code otherwise?